website/project/model.py

```diff
@@ -1267,6 +1267,36 @@ def has_permission_on_children(self, user, permission):
 
         return False
 
+    def find_readable_antecedent(self, auth):
+        """ Returns first antecendant node readable by <user>.
+        """
+
+        next_parent = self.parent_node
+        while next_parent:
+            if next_parent.can_view(auth):
+                return next_parent
+            next_parent = next_parent.parent_node
+
+    def find_readable_descendants(self, auth):
+        """ Returns list of first descendant node readable by <user> in
+        each descendant branch.
+        """
+        ret = []
```

Comment:
In general, it's better for methods like this to be generators rather than returning a list. Leave it up to the caller if they want a list (or some other type of collection).

```python
descendants = set(node.find_readable_descendants(auth)
# or
for node in node.find_readable_descendants(auth):  # we don't load up everything in memory at once
   #..
```